src/MultipartStreamBuilder.php

```diff
@@ -133,28 +138,52 @@ public function addResource($name, $resource, array $options = [])
      */
     public function build()
     {
-        $streams = '';
+        // Assign maximimum 1/4 php's available memory
+        // to attempt buffering the stream content.
+        // If the stream content exceed this, will fallback
+        // to use temporary file.
+        $maxmemory = ($this->bufferMaxMemory < 0)
+            ? \floor(static::getAvailableMemory() / 4)
+            : $this->bufferMaxMemory;
+
+        // Open a temporary read-write stream as buffer.
+        // If the size is less than predefined limit, things will stay in memory.
+        // If the size is more than that, things will be stored in temp file.
+        $buffer = fopen('php://temp/maxmemory:'.$maxmemory, 'r+');
         foreach ($this->data as $data) {
             // Add start and headers
-            $streams .= "--{$this->getBoundary()}\r\n".
-                $this->getHeaders($data['headers'])."\r\n";
+            fwrite($buffer, "--{$this->getBoundary()}\r\n".
+                $this->getHeaders($data['headers'])."\r\n");
 
-            // Convert the stream to string
-            /* @var $contentStream StreamInterface */
+            /**
+             * @var \Psr\Http\Message\StreamInterface
+             */
             $contentStream = $data['contents'];
+
+            // Read stream into buffer
             if ($contentStream->isSeekable()) {
-                $streams .= $contentStream->__toString();
+                $contentStream->rewind(); // rewind to beginning.
+            }
+            if ($contentStream->isReadable()) {
+                while (!$contentStream->eof()) {
+                    // read 8KB chunk into buffer until reached EOF.
+                    fwrite($buffer, $contentStream->read(8192));
```

Comment:
For a large steam, this will consume 8MB, lets reduce it to 1MB chunks instead. You will no lose any performance.

Reply:
This would read 8KB a chunk. I can update it to 1MB.

Reply:
Hehe, Math. =)

Yes please.